src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -0,0 +1,453 @@
+}
+
+void verifyInstallation(const Json::Value& manifest) {
+  // Make sure operation_result and filepath were correctly written and formatted.
+  EXPECT_EQ(manifest["CA:FE:A6:D2:84:9D"]["signed"]["custom"]["operation_result"]["id"].asString(),
+            "primary_firmware.txt");
+  EXPECT_EQ(manifest["CA:FE:A6:D2:84:9D"]["signed"]["custom"]["operation_result"]["result_code"].asInt(),
+            static_cast<int>(data::UpdateResultCode::kOk));
+  EXPECT_EQ(manifest["CA:FE:A6:D2:84:9D"]["signed"]["custom"]["operation_result"]["result_text"].asString(),
+            "Installing fake package was successful");
+  EXPECT_EQ(manifest["CA:FE:A6:D2:84:9D"]["signed"]["installed_image"]["filepath"].asString(), "primary_firmware.txt");
+  // installation_result has not been implemented for secondaries yet.
+  EXPECT_EQ(manifest["secondary_ecu_serial"]["signed"]["installed_image"]["filepath"].asString(),
+            "secondary_firmware.txt");
+}
+
+int num_events_FullNoUpdates = 0;
+void process_events_FullNoUpdates(const std::shared_ptr<event::BaseEvent>& event) {
+  if (event->variant == "DownloadProgressReport") {
+    return;
+  }
+  switch (num_events_FullNoUpdates) {
+    case 0:
+      EXPECT_EQ(event->variant, "FetchMetaComplete");
+      break;
+    case 1:
+      EXPECT_EQ(event->variant, "FetchMetaComplete");
+      break;
+    case 5:
+      // Don't let the test run indefinitely!
+      FAIL();
+    default:
+      std::cout << "event #" << num_events_FullNoUpdates << " is: " << event->variant << "\n";
+      EXPECT_EQ(event->variant, "");
+  }
+  ++num_events_FullNoUpdates;
+}
+
+/*
+ * Using automatic control, test that if there are no updates, no additional
+ * events are generated beyond the expected ones.
+ */
+TEST(Aktualizr, FullNoUpdates) {
+  TemporaryDirectory temp_dir;
+  auto http = std::make_shared<HttpFake>(temp_dir.Path());
+  Config conf = makeTestConfig(temp_dir, http->tls_server);
+  conf.uptane.director_server = http->tls_server + "/noupdates/director";
+  conf.uptane.repo_server = http->tls_server + "/noupdates/repo";
+  conf.uptane.running_mode = RunningMode::kFull;
+
+  auto storage = INvStorage::newStorage(conf.storage);
+  auto sig = std::make_shared<boost::signals2::signal<void(std::shared_ptr<event::BaseEvent>)>>();
+  auto up = SotaUptaneClient::newTestClient(conf, storage, http, sig);
+  Aktualizr aktualizr(conf, storage, up, sig);
+  std::function<void(std::shared_ptr<event::BaseEvent> event)> f_cb = process_events_FullNoUpdates;
+  boost::signals2::connection conn = aktualizr.SetSignalHandler(f_cb);
+
+  EXPECT_NO_THROW(up->initialize());
+  aktualizr.FetchMetadata();
+  // Fetch twice so that we can check for a second FetchMetaComplete and
+  // guarantee that nothing unexpected happened after the first fetch.
+  aktualizr.FetchMetadata();
+
+  size_t counter = 0;
+  while (num_events_FullNoUpdates < 2) {
+    sleep(1);
```

Comment:
> All these polls could be replaced by a wait on future with timeout

Done, although I didn't upgrade the reportqueue tests, because getting the promise into `httpfake` seems like too much of a chore.